clap_complete/src/dynamic/fish.rs

```diff
@@ -0,0 +1,47 @@
+//! Complete commands within fish
+
+// For fish the behviour should match the default as close as possible
+// 1. grouping shorthand completions i.e. -a<TAB> should show other shorthands to end up with
+//    -ams...
+// 2. only complete options when one - is typed
+// Due to https://github.com/fish-shell/fish-shell/issues/7943 we need to implement this our self
+//
+
+use std::ffi::OsString;
+
+use clap::Args;
+
+use super::Completer;
+
+/// Dynamic completion for Fish
+pub struct Fish;
+
+impl Completer for Fish {
+    type CompleteArgs = CompleteArgs;
+
+    fn completion_script(cmd: &mut clap::Command) -> Vec<u8> {
+        let name = cmd.get_bin_name().unwrap_or_else(|| cmd.get_name());
+        format!(
+            r#"complete -x -c {name} -a "{name} complete --commandline (commandline) --cursor (commandline -C)""#
+        ).into_bytes()
+    }
+
+    fn file_name(name: &str) -> String {
+        format!("{name}.fish")
+    }
+
+    fn try_complete(args: &Self::CompleteArgs, cmd: &mut clap::Command) -> clap::error::Result<()> {
+        todo!()
```

Comment:
For fish we need to have a function that returns all possible args at the current position, meaning from the beginning of the last token:
`smth he|ll` => every completion that would be at the same place as `hell` e.g. "hello", "hi",...

And one that returns all the possible options.

This could also be one function if we had a struct representing a completion item that could tell wheather it is an option or an argument

Reply:
As I'm not completely sure what we already have in the bash completions.

@epage do we have functionality that would fit this use case that we could extract into the dynamic mod to use for both bash and fish?